Add Worker Limits to AWS Resource Reference + Improve Worker Docs

paolaperaza opened the pull request:
This PR addresses #627 by doing two things:

- Adds a table in AWS Resource Reference that includes every node instance type we support and a corresponding worker limit maximum, as found by the CRE team
- Improves language in the "Worker Resources" section of our "Configure Deployment" doc

@collinmcnulty @alex-astronomer @stuart23 Would love for you guys to take a look. FYI:

- It's evident that some instance types are too big to provision workers given our 400 AU limit (e.g. `m5d.metal` which has a worker size limit of 967 AU), so I'm going to add a note here that acknowledges that limitation
- It looks like the worker size limit for `t3.xlarge` was not in the spreadsheet. Do you all have that value?

Based on: https://docs.google.com/spreadsheets/d/1-ejFME01oaqI7-n2MfBqVNf1PbR4H1sg4kuui28nERA/edit#gid=0

kwcanuck commented:
@paolaperaza  some feedback:

- While it's acceptable to use _per_ to mean _for each_ in statistical or technical contexts, I would rewrite the section title to avoid the use of _per_. For example: _Node Instance Type Worker Size Limits_ or _Worker Size Limits for Node Types_.

- There are no values for the t3.xlarge instance type. Perhaps this is deliberate. 

- When possible avoid the passive voice (...can be..).  This could be rewritten as: _In addition to setting a node instance type for each Cluster, you can configure a unique worker size for each deployment within a Cluster_.

- Are _Cluster_ and _Deployment_ proper nouns? If not, they shouldn't be capitalized.

- Although ending a sentence with a proposition (with) isn't verboten, I would recommend a rewrite to avoid it.  For example: _…as long as the worker size is supported by the node instance type selected for the Cluster._

- You can use contractions to give the content a more conversational tone. For example, _isn't_ instead of _is not_.

- In general, don't use _we_. Focus on the customer, and avoid making Astronomer the subject. When necessary, use _Astronomer recommends_.

- Avoid idioms such as _out of the box_. They can be confusing for non-native English speakers and are hard to localize.

- Do you want to provide _specific_ timelines in the documentation? For example, _...in the first half of 2022_. What happens when this deadline isn't met? Who will remove this reference after the functionality is available?

- Instead of _read our_ use _see_ instead.

Feel free to implement or ignore my suggestions.

paolaperaza commented:
A sincere thank you @kwcanuck ! I just committed all of your changes, with exception of:

- Adding value for the `t3.xlarge` instance type. Need to get it from CRE
- Kept the "first half of 2022" thing - it's not a great practice and I don't suspect we'll do for much longer but given our scope right now we're pretty good about updating these and they seem to get prospects excited and reaching out to our sales team

Thanks so much again for the review 🙏

collinmcnulty commented on astro/resource-reference-aws.md:
```diff
@@ -91,49 +118,50 @@ For detailed information on each instance type, reference [AWS documentation](ht
+| m5.8xlarge         | 307        | 30.7 CPUs | 115 GiB MEM  |
+| m5.12xlarge        | 467        | 46.7 CPUs | 175 GiB MEM  |
+| m5.16xlarge        | 627        | 62.7 CPUs | 235 GiB MEM  |
+| m5.24xlarge        | 947        | 94.7 CPUs | 355 Gib MEM  |
```
I suggest we say 400 for all node sizes that have maxes above 400 as that is the limit of the slider.

paolaperaza replied:
@collinmcnulty I feel like I'd rather have the max be true here, and then have folks reach out to use if they want to reach this maximum. I have a note below:

> The maximum worker size in the Cloud UI is currently 400 AU. If your team is interested in using an instance type that supports a higher worker size limit, reach out to [Astronomer Support](https://support.astronomer.io).

Think that cuts it or still misleading?

collinmcnulty replied:
Maybe just put an asterisk like 947*?

collinmcnulty commented on astro/resource-reference-aws.md:
```diff
@@ -91,49 +118,50 @@ For detailed information on each instance type, reference [AWS documentation](ht
 
+## Deployment Worker Size Limits
+
+In addition to setting a node instance type for each Cluster, you can configure a unique worker size for each Deployment within a Cluster. Worker size can be specified at any time in the **Worker Resources** field in the Deployment view of the Cloud UI. You can select any worker size up to 400 AU (40 CPUs, 150 GiB memory) as long as the worker size is supported by the node instance type selected  ...
```
Should we also mention scheduler resources? Scheduler can go up to 30 AU which is above the 27 limit for our smallest instances.

paolaperaza replied:
@collinmcnulty Good call, I can add a note at the bottom of this section.

alex-astronomer commented on astro/configure-deployment.md:
```diff
@@ -41,28 +41,35 @@ If you prefer to work with the Astro CLI, you can also create a Deployment by ru
 ### Worker Resources
 
-In Airflow, a worker is responsible for executing tasks that are scheduled and queued by the Scheduler. On Astro, task execution is powered by the [Celery Executor](https://airflow.apache.org/docs/apache-airflow/stable/executor/celery.html) with [KEDA](https://www.astronomer.io/blog/the-keda-autoscaler), which enables a Deployment's worker count to auto-scale between 1 and 10 depending on real-ti ...
+A worker is responsible for executing tasks, which are first scheduled and queued by the Scheduler. On Astro, task execution is powered by the [Celery Executor](https://airflow.apache.org/docs/apache-airflow/stable/executor/celery.html) with [KEDA](https://www.astronomer.io/blog/the-keda-autoscaler), which enables workers to auto-scale between 1 and 10 depending on real-time workload. Each worker ...
```
What limits the workers at 10?  Isn't there some sort of configuration that allows for more than 10 workers?

paolaperaza replied:
It actually is a hard cap at 10 right now: https://github.com/astronomer/airflow-operator/blob/fb8989d99e66047c0d71393f0b0b2341494e170c/config/crd/bases/airflow.apache.org_workers.yaml#L165

Actively working on a new "Min/Max Worker Count" slider though: https://github.com/astronomer/harmony/issues/95

collinmcnulty replied:
Out of scope for this PR, but I feel this is another knob that we could take away from the customer per our conversation yesterday. We already have a control for capping node count.

alex-astronomer reviewed (commented):
Looks good to me, just one question that I have.

paolaperaza commented:
@collinmcnulty @alex-astronomer Thank you both for the feedback! Two final things:

- I added a note about Scheduler resources (the node instance type the cluster selects doesn't actually apply to Scheduler pods)
- I added an `*` to all node instance types with worker limits over 400 AU and a note that full cpu and memory capacity may not be used

**To finish this up, could either of you add the following new instance types to the spreadsheet with a calculated max AU?**

- c6i
c6i.xlarge
c6i.2xlarge
c6i.4xlarge
c6i.8xlarge
c6i.12xlarge
c6i.16xlarge
c6i.24xlarge
c6i.32xlarge
c6i.metal

- m6i
m6i.xlarge
m6i.2xlarge
m6i.4xlarge
m6i.8xlarge
m6i.12xlarge
m6i.16xlarge
m6i.24xlarge
m6i.32xlarge
m6i.metal

- r6i
r6i.xlarge
r6i.2xlarge
r6i.4xlarge
r6i.8xlarge
r6i.12xlarge
r6i.16xlarge
r6i.24xlarge
r6i.32xlarge
r6i.metal

These new instance types came out end of last week! I know some of these are CPU/Memory optimized, so the AU numbers might be wonky. It is what it is, we'll optimize here later

collinmcnulty commented:
The m6i and r6i series are going to have exactly the same limits as the M5 series. C6 series is going to require a new formula, as it's likely memory-limited while all the others are CPU limited.

collinmcnulty commented:
Looking at the daemon sets, I'm not clear now that they are CPU limited. You won’t steer anyone into a broken by using the m5 numbers for all of them. We can do more testing tomorrow.

paolaperaza commented on astro/configure-deployment.md:
```diff
@@ -41,28 +41,35 @@ If you prefer to work with the Astro CLI, you can also create a Deployment by ru
 ### Worker Resources
 
-In Airflow, a worker is responsible for executing tasks that are scheduled and queued by the Scheduler. On Astro, task execution is powered by the [Celery Executor](https://airflow.apache.org/docs/apache-airflow/stable/executor/celery.html) with [KEDA](https://www.astronomer.io/blog/the-keda-autoscaler), which enables a Deployment's worker count to auto-scale between 1 and 10 depending on real-ti ...
+A worker is responsible for executing tasks, which are first scheduled and queued by the Scheduler. On Astro, task execution is powered by the [Celery Executor](https://airflow.apache.org/docs/apache-airflow/stable/executor/celery.html) with [KEDA](https://www.astronomer.io/blog/the-keda-autoscaler), which enables workers to auto-scale between 1 and 10 depending on real-time workload. Each worker ...
```
```suggestion
A worker is responsible for executing tasks, which are first scheduled and queued by the Scheduler. On Astro, task execution is powered by the [Celery Executor](https://airflow.apache.org/docs/apache-airflow/stable/executor/celery.html) with [KEDA](https://www.astronomer.io/blog/the-keda-autoscaler), which enables workers to autoscale between 1 and 10 depending on real-time workload. Each worker is a Kubernetes Pod that is hosted within a Kubernetes Node in your Astro Cluster.
```

paolaperaza commented on astro/configure-deployment.md:
```diff
@@ -41,28 +41,35 @@ If you prefer to work with the Astro CLI, you can also create a Deployment by ru
+To modify the resources allocated to the workers in your Deployment, specify a quantity of AUs in the **Worker Resources** field of the Cloud UI. This value determines the size of each worker in your Deployment. While the total number of workers running at any given time may change, all workers will be created with the amount of CPU and memory you specify here. If you set **Worker Resources** to  ...
 
-All Celery workers assume the same resources. If you set **Worker Resources** to 10 AU, for example, your Deployment might scale up to 3 workers using 10 AU each for a total of 30 AU (3 CPU, 11.25 GB Memory). By default, the minimum AU allocated towards workers is 10.
+To ensure reliability, the minimum worker size supported is 10 AU. Beyond that, the maximum worker size you can set depends on the node instance type that is configured for the Cluster in which your Deployment is hosted. If you attempt to provision a worker size that is not supported by your Cluster's instance type, you will see an error in the Cloud UI. For example, if the node instance type for ...
```
```suggestion
To ensure reliability, the minimum worker size supported is 10 AU. Beyond that, the maximum worker size you can set depends on the node instance type that is configured for the Cluster in which your Deployment is hosted. If you attempt to provision a worker size that is not supported by your Cluster's instance type, you will see an error in the Cloud UI. For example, if the node instance type for a given Cluster is set to `m5.xlarge`, the maximum worker size supported for any Deployment within that Cluster is 27 AU (2.7 CPUs, 10.1 GiB memory). This limit accounts for overhead that is required for system components.
```

paolaperaza commented on astro/configure-deployment.md:
```diff
@@ -41,28 +41,35 @@ If you prefer to work with the Astro CLI, you can also create a Deployment by ru
+To ensure reliability, the minimum worker size supported is 10 AU. Beyond that, the maximum worker size you can set depends on the node instance type that is configured for the Cluster in which your Deployment is hosted. If you attempt to provision a worker size that is not supported by your Cluster's instance type, you will see an error in the Cloud UI. For example, if the node instance type for ...
 
-The ability to set minimum and/or maximum number of workers is coming soon.
+For a list of node instance types that are supported on Astro and their corresponding worker size limits, see [AWS Resource Reference](aws-resource-reference.md#deployment-worker-size-limits). To request a different instance type for your Cluster, reach out to [Astronomer Support](https://support.astronomer.io).
```
```suggestion
For a list of supported node instance types and their corresponding worker size limits, see [AWS Resource Reference](aws-resource-reference.md#deployment-worker-size-limits). To request a different instance type for your Cluster, reach out to [Astronomer Support](https://support.astronomer.io).
```
